satellite/satellitedb/orders.go

```diff
@@ -266,18 +266,31 @@ func (db *ordersDB) ProcessOrders(ctx context.Context, requests []*orders.Proces
 	// the case where the order has already been processed.  Duplicates and previously
 	// processed orders are rejected
 	for _, request := range requests {
-		insert := "INSERT INTO used_serials (serial_number_id, storage_node_id) SELECT id, ? FROM serial_numbers WHERE serial_number = ?"
+		// create a savepoint so we can rollback if an order has already been used
+		_, err = tx.Exec("savepoint sp;")
 		if err != nil {
 			return nil, err
 		}
+
+		insert := "INSERT INTO used_serials (serial_number_id, storage_node_id) SELECT id, ? FROM serial_numbers WHERE serial_number = ?"
+
 		_, err = tx.Exec(db.db.Rebind(insert), request.OrderLimit.StorageNodeId.Bytes(), request.OrderLimit.SerialNumber.Bytes())
 		if err != nil {
 			if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
 				reject(request.OrderLimit.SerialNumber)
+				// rollback to the savepoint before the insert failed
+				_, err = tx.Exec("rollback to savepoint sp;")
```

Comment:
~I think that what it solves the issue isn't the `savepoint` is that in this conditional the `err` should have been assigned to `nil` before the next iteration.~

~Probably what's happening now is that `err` isn't checked anymore in the next iteration because of the added call to `_, err = tx.Exec("savepoint sp;")`.~

~Anyway, it looks to me that the `if err != nil` conditional of line 270 (previous changes) wasn't needed and here a `continue` statement should have been called.~